reference-implementation/examples/test_vectors.rs

```diff
@@ -0,0 +1,313 @@
+    use sframe_reference::header::*;
+
+    const MARKDOWN_SNIPPET: &str = "~~~
+KID: 0x{kid}
+CTR: 0x{ctr}
+Header: {header}
+~~~";
+
+    #[derive(Serialize)]
+    pub struct TestVector {
+        kid: u64,
+        ctr: u64,
+        encoded: String,
+    }
+
+    impl TestVector {
+        fn new(kid: u64, ctr: u64) -> Self {
+            let header = Header::new(KeyId(kid), Counter(ctr));
+            let encoded = hex::encode(header.to_vec());
+            Self { kid, ctr, encoded }
+        }
+
+        pub fn make_all() -> Vec<Self> {
+            let mut values: Vec<u64> = (0..64)
+                .step_by(8)
+                .map(|n| [(1 << n) - 1, (1 << n)])
+                .flatten()
+                .collect();
+            values.push(u64::MAX);
+
+            let kids = values.clone();
+            let ctrs = values.clone();
+
+            kids.into_iter()
+                .cartesian_product(ctrs.into_iter())
+                .map(|(kid, ctr)| Self::new(kid, ctr))
+                .collect()
+        }
+    }
+
+    impl super::ToMarkdown for TestVector {
+        fn to_markdown(&self) -> String {
+            MARKDOWN_SNIPPET
+                .replace("{kid}", &format!("{:016x}", self.kid))
+                .replace("{ctr}", &format!("{:016x}", self.ctr))
+                .replace("{header}", &self.encoded)
+        }
+    }
+}
+
+mod aes_ctr_hmac {
+    use aead::{Aead, Key, KeyInit, KeySizeUser, Nonce, Payload};
+    use aes::Aes128;
+    use cipher::consts::{U10, U16, U4, U8};
+    use cipher::ArrayLength;
+    use hex_literal::hex;
+    use serde::Serialize;
+    use sframe_reference::aes_ctr_hmac::*;
+    use sframe_reference::cipher::CipherSuite;
+    use sha2::Sha256;
+
+    const MARKDOWN_SNIPPET: &str = "~~~
+Cipher suite: {cipher_suite}
+`key`: {key}
+`aead_label`: {aead_label}
+`aead_secret`: {aead_secret}
+`enc_key`: {enc_key}
+`auth_key`: {auth_key}
+`nonce`: {nonce}
+`aad`: {aad}
+`pt`: {pt}
+`ct`: {ct}
+~~~";
+
+    #[derive(Serialize)]
+    pub struct TestVector {
+        cipher_suite: u16,
+        key: String,
+        aead_label: String,
+        aead_secret: String,
+        enc_key: String,
+        auth_key: String,
+        nonce: String,
+        aad: String,
+        pt: String,
+        ct: String,
+    }
+
+    impl TestVector {
+        fn new<C, D, T>() -> Self
+        where
+            C: Cipher + KeySizeUser<KeySize = U16>,
+            D: Digest,
+            T: ArrayLength<u8>,
+        {
+            let cipher_suite = match T::to_usize() {
+                10 => CipherSuite::AES_128_CTR_HMAC_SHA_256_80,
+                8 => CipherSuite::AES_128_CTR_HMAC_SHA_256_64,
+                4 => CipherSuite::AES_128_CTR_HMAC_SHA_256_32,
+                _ => unreachable!(),
+            };
+
+            let key: Key<AesCtrHmac<C, D, T>> = hex!("000102030405060708090a0b0c0d0e0f").into();
+            let nonce: Nonce<AesCtrHmac<C, D, T>> = hex!("101112131415161718191a1b").into();
```

Comment:
Should this be `16` bytes for the CTR mode instead of `12`?

Reply:
I was thinking we would use a 12-byte nonce, so that the resulting 16-byte IV would have the right-most 4 bytes set to zero for the counter. (Similar to [what SRTP does](https://datatracker.ietf.org/doc/html/rfc3711#section-4.1.1), though SRTP only has two zero bytes.)  This aligns well with the nonce formation scheme (`nonce = xor(sframe_salt, CTR)`), since adding the four counter bytes after the nonce formation means that the counter and the nonce don't interfere with each other.

We probably need a better explanation of this in the spec, though.  Filed #132

Reply:
Thanks, that's clearer. Padding with zeros sounds totally fine to me, but a better explanation would be nice.

Reply:
Note that this padding happens in `nonce_to_iv` in this implementation.